### PR TITLE
Feature/classic

### PR DESCRIPTION
**Pull request by dburkhardt**

### CODE REVIEW

File: python/magic/version.py [hunk 1]
```diff
@@ -1 +1 @@
-__version__ = "1.5.7"
+__version__ = "1.5.8a0"
```
Comment (scottgigante) [inline python/magic/version.py]: this should be 1.5.8

File: python/test/test.py [hunk 2]
```diff
@@ -22,7 +22,7 @@ def test_scdata():
     scdata_norm = scprep.transform.sqrt(scdata_norm)
     assert scdata.shape == scdata_norm.shape
     np.random.seed(42)
-    magic_op = magic.MAGIC(t='auto', a=20, k=10)
+    magic_op = magic.MAGIC(t='auto', a=20, k=10, verbose=False)
```
Comment (scottgigante) [inline python/test/test.py]: Can you replace the `k` with `knn` and `a` with `decay`?

Comment (scottgigante) [reply python/test/test.py]: this goes for all of the tests

File: python/test/test.py [hunk 3]
```diff
@@ -39,6 +39,14 @@ def test_scdata():
     dremi = magic_op.knnDREMI("VIM", "ZEB1", plot=True)
     np.testing.assert_allclose(dremi, 1.573619, atol=0.0000005)
 
+    # Testing exact vs approximate solver
+    magic_op = magic.MAGIC(t='auto', a=20, k=10, solver='exact', verbose=False)
+    data_imputed = magic_op.fit_transform(scdata_norm)
+    assert np.all(data_imputed >= 0)
+
+    magic_op.set_params(solver='approximate')
+    magic_op.fit_transform(scdata_norm)
```
Comment (scottgigante) [inline python/test/test.py]: Can we add a test here to check that there _are_ negative values?

File: python/magic/version.py (hunk 1 above)
Comment (dburkhardt) [reply python/magic/version.py]: Done

File: python/magic/magic.py [hunk 4]
```diff
@@ -647,8 +659,13 @@ def _impute(self, data, t_max=20, plot=False, ax=None,
...
         """
-        if not isinstance(data, graphtools.base.Data):
-            data = graphtools.base.Data(data, n_pca=self.n_pca)
+
+        if self.solver == 'approximate':
+            if not isinstance(data, graphtools.base.Data):
+                data = graphtools.base.Data(data, n_pca=self.n_pca)
+        elif self.solver == ' exact':
+            if not isinstance(data, graphtools.base.Data):
+                data = graphtools.base.Data(data, n_pca=None)
```
Comment (scottgigante) [inline python/magic/magic.py]: This can be simplified to
```
if not isinstance(data, graphtools.base.Data):
    data = graphtools.base.Data(data, n_pca=self.n_pca if self.solver == 'approximate' else None)
```

File: python/magic/magic.py [hunk 5]
```diff
@@ -157,7 +163,7 @@ def __init__(self, knn=10, decay=15, t='auto', n_pca=100,
         self.knn_dist = knn_dist
         self.n_jobs = n_jobs
         self.random_state = random_state
-
+        self.solver = solver
```
Comment (scottgigante) [inline python/magic/magic.py]: Can you add a check to the effect of `assert self.solver in ['exact', 'approximate']`? But use a ValueError with a nice error message

Comment (scottgigante) [reply python/magic/magic.py]: nvm that's on line 208

File: python/magic/magic.py (hunk 4 above)
Comment (dburkhardt) [reply python/magic/magic.py]: Yes, but that makes the code control flow less clear. See GVM (https://blog.dropbox.com/topics/company/thank-you--guido)

>>>"My opinion [is] that maintainable code is more important than clever code... If I encountered clever code that was particularly cryptic, and I had to do some maintenance on it, I would probably rewrite it

File: python/test/test.py (hunk 3 above)
Comment (dburkhardt) [reply python/test/test.py]: Ugh fine you were right

**Comment by dburkhardt** [comment]
@scottgigante I think I got all your changes

File: python/magic/magic.py (hunk 4 above)
Comment (scottgigante) [reply python/magic/magic.py]: Okay, but then I'll still argue for less duplication of code
```

        if not isinstance(data, graphtools.base.Data):
            if self.solver == 'approximate':
                data = graphtools.base.Data(data, n_pca=self.n_pca)
            elif self.solver == 'exact':
                data = graphtools.base.Data(data, n_pca=None)
```

Comment (dburkhardt) [reply python/magic/magic.py]: Touché
